File: frappe_k8s_manager/public/js/subscription_dashboard.js
frappe.pages['subscription-dashboard'].on_page_load = function(wrapper) {
    let page = frappe.ui.make_app_page({
        parent: wrapper,
        title: 'Subscription Dashboard',
        single_column: true
    });

    frappe.call({
        method: 'frappe_k8s_manager.www.subscription_dashboard.get_subscription',
        args: { user_email: frappe.session.user },
        callback: function(response) {
            let data = response.message;
            $(wrapper).html(`
                <div class="subscription-info">
                    <h3>Current Plan: ${data.plan}</h3>
                    <p>Disk Space: ${data.disk_space} GB</p>
                    <p>Installed Apps: ${data.apps}</p>
                    <p>Status: ${data.status}</p>
                    <button class="btn btn-primary" id="upgrade-plan">Upgrade</button>
                    <button class="btn btn-warning" id="downgrade-plan">Downgrade</button>
                    <button class="btn btn-danger" id="cancel-subscription">Cancel</button>
                </div>
            `);

            $('#upgrade-plan').click(() => updateSubscription('upgrade'));
            $('#downgrade-plan').click(() => updateSubscription('downgrade'));
            $('#cancel-subscription').click(() => updateSubscription('cancel'));
        }
    });

    function updateSubscription(action) {
        let new_plan = action !== "cancel" ? prompt("Enter New Plan ID") : null;
        frappe.call({
            method: 'frappe_k8s_manager.www.subscription_dashboard.update_subscription',
            args: { user_email: frappe.session.user, action: action, new_plan_id: new_plan },
            callback: function(response) {
                frappe.msgprint(response.message);
            }
        });
    }
};

File: frappe_k8s_manager/www/subscription_dashboard.py
import frappe
from frappe_k8s_manager.kubernetes.manage import manage_subscription

@frappe.whitelist()
def get_subscription(user_email):
    """Fetch user subscription details."""
    try:
        user = frappe.get_doc("User", {"email": user_email})
        subscription = frappe.get_doc("Subscription", {"user": user.name})

        return {
            "success": True,
            "plan": subscription.plan,
            "disk_space": subscription.disk_space,
            "apps": subscription.apps,
            "status": subscription.status
        }
    except frappe.DoesNotExistError:
        return {"success": False, "error": "User or subscription not found"}
    except Exception as e:
        frappe.log_error(f"Error fetching subscription for {user_email}: {str(e)}")
        return {"success": False, "error": "An unexpected error occurred"}

@frappe.whitelist()
def update_subscription(user_email, action, new_plan_id=None):
    """Upgrade, downgrade, or cancel the subscription."""
    try:
        # Validate input
        if action in ["upgrade", "downgrade"] and not new_plan_id:
            return {"success": False, "error": "New plan ID is required for this action"}

        result = manage_subscription(action, user_email, new_plan_id)

        return {"success": True, "message": "Subscription updated successfully", "details": result}
    except frappe.DoesNotExistError:
        return {"success": False, "error": "User or subscription not found"}
    except Exception as e:
        frappe.log_error(f"Error updating subscription for {user_email}: {str(e)}")
        return {"success": False, "error": "An unexpected error occurred"}
